anti-evil-maid/90anti-evil-maid/anti-evil-maid-unseal

```diff
@@ -134,71 +131,180 @@ else
              break
         fi
 
-        log "Wrong SRK password, resetting dictionary attack lock..."
-        tpm_resetdalock -z || true  # -z is an empty owner (not SRK) password
+        log "Wrong SRK password"
     done
 fi
 
-log "Unsealing the secret..."
-if tpm_unsealdata $Z -i "$SEALED_SECRET" -o "$UNSEALED_SECRET" \
-   <"$SRK_PASSWORD_CACHE"; then
-    rm -rf "$CACHE_DIR"
+
+# check freshness token
+
+log "Unsealing freshness token..."
+if tpm_unsealdata $Z -i "$sealed_secret_fsh" -o /tmp/fresh \
+        < "$SRK_PASSWORD_CACHE"; then
+    log "Freshness token unsealed."
+else
+    log "Freshness token unsealing failed!"
+    exit 1
 fi
 
-log "Unmounting the $LABEL device..."
-umount "$MNT"
+if checkfreshness /tmp/fresh; then
+    log "Freshness token valid, continuing."
+else
+    log "Freshness token invalid!"
+    exit 1
+fi
+
+rm -f /tmp/fresh
+
 
+# check for OTP
 
-if [ ! -d "$CACHE_DIR" ]; then  # unseal succeeded
-    if png; then
-        # display secret in next dialog
-        WHERE="next to the prompt for it"
-        if file "$UNSEALED_SECRET" 2>/dev/null | grep -q PNG; then
-            cp "$UNSEALED_SECRET" "$PLYMOUTH_THEME_UNSEALED_SECRET"
+if [ -e "$sealed_secret_otp" ]; then
+    alias otp=true
+    SEALED_SECRET=$sealed_secret_otp
+else
+    alias otp=false
+fi
+
+
+# tell user to remove AEM dev if RO
+
+message ""
+message "You should now unplug the AEM device if it's read-only."
+message ""
+
+
+# unseal & show OTP if provisioned
+
+if otp; then
+    log "Unsealing TOTP shared secret seed..."
+    if tpm_unsealdata $Z -i "$SEALED_SECRET" -o "$UNSEALED_SECRET" \
+            < "$SRK_PASSWORD_CACHE"; then
+        log "TOTP secret unsealed."
+        seed=$(cat "$UNSEALED_SECRET")
+
+        message "Please verify your OTP device displays the same TOTP code"
+        message "as shown below. Should the codes match, press <ENTER> to"
+        message "unseal your LUKS key file. If you do now have access to"
+        message "your OTP device but wish to continue despite the security"
+        message "implications, press <ENTER>. Pressing <Q> will revert"
+        message "back to displaying the text secret (if provisioned)."
+        message "Otherwise, shut down the computer."
+        message ""
+        message "[ $(date) ] TOTP code: $(oathtool --totp -b "$seed")"
+        last=$(date +%s)
+
+        if plymouth_active; then
+            plymouth watch-keystroke --keys=$'\n' --command='touch /tmp/mfa-cont' &
+            plymouth watch-keystroke --keys=Qq --command='touch /tmp/mfa-skip' &
+            while [ ! -e /tmp/mfa-cont ] && [ ! -e /tmp/mfa-skip ]; do
```

Comment:
It's reasonable for this loop to run longer than systemd's `DefaultTimeoutStartSec` (90 seconds), so `TimeoutStartSec` should probably be bumped to something like `5m` in `-unseal.service`